cpp/include/raft/sparse/sparse_matrix.hpp

```diff
@@ -0,0 +1,269 @@
+ * @param [in] vdim total number of dimensions (e.g. cols for csr, rows for csc)
+ * @param [in] src pointer of source indices for each edge (size nnz)
+ * @param [in] dst pointer of destination indices for each edge (size nnz)
+ * @param [in] values optional pointer of values or weights for each edge (size nnz)
+ * @param [in] nnz number of nonzero values or edges
+ *
+ */
+template <typename src_idx_t, typename dst_idx_t, typename value_t>
+class coo_device_view_t {
+ public:
+  // TODO: Use a strategy / policy to separate host from device views
+  coo_view_t(src_idx_t nvec,         // total number of source vectors
+                    dst_idx_t vdim,  // total number of destination vectors
+                    array_view_t<src_idx_t> const src,
+                    array_view_t<dst_idx_t> const dst,
+                    std::optional<array_view_t<value_t> const> values,
+                    size_t nnz)
+    : src_(src),
+      dst_(dst),
+      values_(values),
+      nvec_(nvec),
+      vdim_(vdim),
+      nnz_(nnz)
+  {
+  }
+
+  /**
+   * Using blas terminology here to abstract away the notion of graph
+   */
+    _RAFT_SPARSE_HD src_idx_t nvec() const { return nvec_; }
+    _RAFT_SPARSE_HD dst_idx_t vdim() const { return vdim_; }
+
+    _RAFT_SPARSE_HD size_t nnz() const { return nnz_; }
+
+    _RAFT_SPARSE_HD array_view_t<src_idx_t> const src() const { return src_; }
+    _RAFT_SPARSE_HD array_view_t<dst_idx_t> const dst() const { return dst_; }
+    _RAFT_SPARSE_HD std::optional<array_view_t<value_t> const> values() const { return values_; }
+
+ private:
+  // TODO: Use span/mdspan here
+  src_idx_t nvec_{0};
+  dst_idx_t vdim_{0};
+  array_view_t<src_idx_t> const src_;
+  array_view_t<dst_idx_t> const dst_{nullptr};
+  std::optional<array_view_t<value_t> const> values_{std::nullopt};
+  offset_t nnz_{0};
+};
+
+
+
+/**
+ * Compressed sparse object which can be used to represent an adjacency list for both
+ * graph and bigraph in (b)csr/csc formats. The distinction here is made very clear-
+ * since in libraries like cuGraph, the matrix is almost always assumed to be square
+ * (e.g. nvec == vdim) but in many uses in cuML, the matrix might not be square.
+ *
+ * This object can be used to represent structural associations (e.g. no values array)
+ * or weighed / valued associations.
+ *
+ * @tparam idx_t
+ * @tparam offset_t
+ * @tparam value_t
+ *
+ * @param [in] nvec total number of vectors for axis==0 (e.g. rows for csr, cols for csc)
+ * @param [in] vdim total number of dimensions (e.g. cols for csr, rows for csc)
+ * @param [in] offsets pointer of offsets into indices array for each vector idx
+ * @param [in] indices pointer of indices from vector offsets (size nnz)
+ * @param [in] values optional pointer of values for each vector offset (size nnz)
+ * @param [in] nnz number of nonzero offsets
+ *
+ */
+template <typename idx_t, typename offset_t, typename value_t, typename ContainerPolicy>
+class comp_sparse_t {
```

Comment:
Would you consider adding a way to get a sparse matrix view directly out of this sparse matrix container, for example by conversion operator, `.to_view()` method, or `to_view` customization point?

This conversion would encourage developers to write all sparse matrix algorithms as functions taking a sparse matrix view.  Otherwise, developers might be tempted to write overloads that take containers.  This would unnecessarily add to build times, and would hinder writing self-documenting code.  (If a function views a sparse matrix, then the function should take the sparse matrix as a view type.  By analogy, the intent of `void f(std::span<float>)` is much more focused and clear than the intent of `void f(std::vector<float>&)`.  Without documentation, I have to assume that the latter could invalidate iterators.  `f(std::span<float>)` _cannot_ invalidate iterators.)